tests/Validation/ValidationTest.php

```diff
@@ -0,0 +1,196 @@
+                if (strpos(\strrev((string)$file), \strrev(".php")) === 0 && !\in_array(basename((string)$file), $skipped)) {
+                    if ($file->getSize() < 100000) {
+                        $testProviderArray[$frameworkName . "::" . $file->getBasename()] = [$file->getPathname(), $frameworkName];
+                    }
+                }
+            }
+        }
+
+        if (count($testProviderArray) === 0) {
+            throw new Exception("ERROR: Validation testsuite frameworks not found - run `git submodule update --init --recursive` to download.");
+        }
+
+        return $testProviderArray;
+    }
+
+    /**
+     * @group validation
+     * @dataProvider frameworkErrorProvider
+     * @param $testCaseFile
+     * @param $frameworkName
+     */
+    public function testDefinitionErrors($testCaseFile, $frameworkName)
+    {
+        $fileContents = file_get_contents($testCaseFile);
+        $actualValues = $this->getActualTestValues($testCaseFile, $fileContents);
+
+        $outputFile = getExpectedValuesFile($testCaseFile);
+        if (!file_exists($outputFile)) {
+            file_put_contents($outputFile, json_encode($actualValues, JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES));
+        }
+
+        $expectedValues = (array)json_decode(file_get_contents($outputFile));
+
+        try {
+            $this->assertEquals($expectedValues['definitions'], $actualValues['definitions']);
+
+            try {
+                $this->assertArraySubset((array)$expectedValues['references'], (array)$actualValues['references'], false, 'references don\'t match.');
+            } catch (\Throwable $e) {
+                $this->assertEquals((array)$expectedValues['references'], (array)$actualValues['references'], 'references don\'t match.');
+            }
+        } catch (\Throwable $e) {
+            $outputFile = getExpectedValuesFile($testCaseFile);
+            if ($frameworkName === '_cases') {
+                file_put_contents($outputFile, json_encode($actualValues, JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES));
+            }
+
+            throw $e;
+        }
+    }
+
+    private function getActualTestValues($filename, $fileContents): array
+    {
+        global $frameworksDir;
+
+        $index = new Index();
+        $parser = new PhpParser\Parser();
+        $docBlockFactory = DocBlockFactory::createInstance();
+        $definitionResolver = new DefinitionResolver($index);
+
+        $document = new PhpDocument($filename, $fileContents, $index, $parser, $docBlockFactory, $definitionResolver);
+
+        $actualRefs = $index->references;
+        $this->filterSkippedReferences($actualRefs);
+        $actualDefs = $this->getTestValuesFromDefs($document->getDefinitions());
+
+        // TODO - there's probably a more PHP-typical way to do this. Need to compare the objects parsed from json files
+        // to the real results. json_decode returns stdClass Objects, not arrays.
```

Comment:
`json_decode` returns associative arrays if you set the second parameter to `true`.
`==` and `assertEquals` return true if objects are of the same class and have the same properties and arrays have the same exact members at the same indices, recursively.